Library/Formula/mlton.rb

```diff
@@ -0,0 +1,131 @@
+
+# MLton is a self-hosting compiler for Standard ML.
+# In order to build from source, this formula uses the corresponding
+# upstream binary release to bootstrap.
+
+class Mlton < Formula
+  desc "Whole-program, optimizing compiler for Standard ML"
+  homepage "http://mlton.org"
+  url "https://downloads.sourceforge.net/project/mlton/mlton/20130715/mlton-20130715.src.tgz"
+  version "20130715"
+  sha256 "215857ad11d44f8d94c27f75e74017aa44b2c9703304bcec9e38c20433143d6c"
+
+  depends_on "gmp"
+
+  # The corresponding upstream binary release used to bootstrap.
+  resource "bootstrap" do
+    url "https://downloads.sourceforge.net/project/mlton/mlton/20130715/mlton-20130715-2.amd64-darwin.gmp-static.tgz"
+    sha256 "16a6d4e300f45f4af094692cf8033390e4634fa4c072caf6e9c288234100ad22"
+  end
+
+  # The upstream source release hard codes common locations for GMP;
+  # on darwin, these include '/opt/local' (MacPorts) and '/sw' (Fink).
+  # Replace these with placeholders for the Homebrew GMP location.
+  patch :DATA
```

Comment:
Requiring the user to manually edit the script to point to the installation location sounds like a bug worth reporting upstream. That info (`PREFIX`) is available when `make` is called, and the change could be done automatically as part of the `script` target in the `Makefile`, where it produces the `mlton` script.

```
script:
	$(CP) bin/mlton-script "$(MLTON)"
	chmod a+x "$(MLTON)"
	$(CP) "$(SRC)/bin/platform" "$(LIB)"
...
```

In fact, it seems like all or most of the hardcoded `/usr/local` paths in `mlton-script` (and maybe `bin/add-cross` and `bin/regression` too) should be replaced with `PREFIX` at `make` time.

The GMP location, if it's allowed to vary from `PREFIX`, could be parameterized as a Makefile variable, which could preserve the existing behavior and let Homebrew or other custom installations repoint it using a `make` argument, instead of requiring a patch.

Regardless, if we're repointing via either patches or a `make` variable, the Homebrew locations should maybe be pointing to `brew --prefix` instead of `Formula["gmp"].opt_prefix`. The `gmp` formula is not keg-only, so its `opt_prefix` isn't specifically required – it'll be linked in to `$(brew --prefix)/lib` and so on. And I think you'd want to have the other libraries installed in this prefix available too. That would be more like the default MLton behavior, which pulls in everything in `/usr/local`.

(If you end up submitting upstream, you might also ask about this section in the Makefile.
```
PREFIX := /usr
ifeq ($(findstring $(TARGET_OS), darwin freebsd solaris), $(TARGET_OS))
PREFIX := /usr/local
endif
```
It seems odd to make the `/usr` vs `/usr/local` distinction on the basis of OS version; seems like it should be `/usr/local` everywhere that doesn't have a special platform-dependent location like `/mingw`.)